elide-core/src/main/java/com/yahoo/elide/security/PermissionManager.java

```diff
@@ -0,0 +1,278 @@
+
+        boolean shouldDefer = CreatePermission.class.equals(annotationClass)
+                || UpdatePermission.class.equals(annotationClass);
+
+        if (!shouldDefer) {
+            checkFn = (check, changespec) -> check.ok(resource.getRequestScope(), changespec)
+                    && check.ok(resource.getObject(), resource.getRequestScope(), changespec);
+        }
+
+        executePermissions(checks,
+                isAny ? CheckMode.ANY : CheckMode.ALL,
+                resource,
+                Optional.ofNullable(changeSpec),
+                shouldDefer,
+                isFieldAware,
+                checkFn);
+    }
+
+    /**
+     * Execute a set of permission checks.
+     *
+     * @param checks Array of Check annotations
+     * @param mode true if ANY, else ALL
+     * @param resource provided PersistentResource to check
+     * @param changeSpec specification of changes
+     * @param shouldDefer Determine whether or not this check should be re-run in its deferred form (i.e. commitCheck)
+     * @param isFieldAware Determine whether or not this is a field-aware check
+     * @param checkFn the function that actually executes the check so this logic can be reused
+     */
+    private void executePermissions(Class<? extends Check>[] checks,
+                                    CheckMode mode,
+                                    PersistentResource<?> resource,
+                                    Optional<ChangeSpec<?>> changeSpec,
+                                    boolean shouldDefer,
+                                    boolean isFieldAware,
+                                    BiFunction<Check, Optional<ChangeSpec<?>>, Boolean> checkFn) {
+        // TODO: Need to write field-aware checks.
+        for (Class<? extends Check> check : checks) {
+            Check checkHandler;
+            try {
+                checkHandler = check.newInstance();
+            } catch (InstantiationException | IllegalAccessException e) {
+                throw new InvalidSyntaxException("Illegal permission check " + check.getName(), e);
+            }
+
+            Boolean ok = checkFn.apply(checkHandler, changeSpec);
+
+            if (ok == null) {
+                ok = false;
+            }
+
+            if (ok && mode == CheckMode.ANY && !shouldDefer) {
+                return;
+            }
+
+            if (!ok && mode == CheckMode.ALL) {
+                log.debug("ForbiddenAccess {} {}#{}", check, resource.getType(), resource.getId());
+                throw new ForbiddenAccessException();
+            }
+
+            // Add check to a list for later execution
+            if (shouldDefer) {
+                // Hrm so much reliance on side-effects :(
+                // TODO: Hrm. If hashes don't match, this will probably be run far too many times...
+                commitChecks.add(() -> {
+                    executePermissions(checks,
```

Comment:
This is not exactly true (there are two code paths in the function: `shouldDefer` and `!shouldDefer`), but it is possible to have repeats of the same commit checks depending on how lambdas are hashed.

For instance, this will certainly never cause the same `operation` check to be executed multiple times.